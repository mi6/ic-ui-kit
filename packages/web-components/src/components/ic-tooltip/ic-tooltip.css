@import "../../global/normalize.css";

/**
 * @prop --ic-z-index-tooltip: z-index of tooltip
 */

:host(.ic-tooltip) .ic-tooltip-container {
  background-color: var(--ic-architectural-800);
  color: #ffff;
  text-align: center;
  padding: var(--ic-space-xxxs) 0.625rem;
  border-radius: var(--ic-border-radius);
  border: var(--ic-space-1px) solid var(--ic-architectural-40);
  position: absolute;
  max-width: 20rem;
  display: none;
  z-index: var(--ic-z-index-tooltip);
  box-shadow: var(--ic-elevation-overlay);
}

:host(.tooltip-navigation-item:not(.tooltip-navigation-item-side-nav-collapsed))
  .ic-tooltip-container,
:host(.tooltip-disabled) .ic-tooltip-container {
  display: none !important;
}

:host(.tooltip-long-label-navigation-item-side-nav-expanded)
  .ic-tooltip-container[data-show] {
  display: block !important;
}

:host(.ic-tooltip) .ic-tooltip-arrow,
:host(.ic-tooltip) .ic-tooltip-arrow::before {
  position: absolute;
  background: inherit;
}

:host(.ic-tooltip) .ic-tooltip-arrow {
  visibility: hidden;
  width: var(--ic-space-md);
  height: var(--ic-space-md);
}

:host(.ic-tooltip) .ic-tooltip-arrow::before {
  visibility: visible;
  content: "";
  border: var(--ic-space-1px) solid var(--ic-architectural-40);
  width: 100%;
  height: 100%;
}

:host(.ic-tooltip) .ic-tooltip-container[data-show] {
  display: block;
}

:host(.ic-tooltip)
  .ic-tooltip-container[data-popper-placement^="top"]
  > .ic-tooltip-arrow {
  height: var(--ic-space-xxxs);
}

:host(.ic-tooltip)
  .ic-tooltip-container[data-popper-placement^="top"]
  > .ic-tooltip-arrow::before {
  border-radius: 0 0 var(--ic-border-radius) var(--ic-border-radius);
  left: 0;
  top: var(--ic-space-xxxs);
  border-top: 0;
}

:host(.ic-tooltip)
  .ic-tooltip-container[data-popper-placement^="bottom"]
  > .ic-tooltip-arrow {
  top: calc(-1 * var(--ic-space-xxs));
  height: var(--ic-space-xxxs);
}

:host(.ic-tooltip)
  .ic-tooltip-container[data-popper-placement^="bottom"]
  > .ic-tooltip-arrow::before {
  border-radius: var(--ic-border-radius) var(--ic-border-radius) 0 0;
  left: 0;
  top: var(--ic-space-1px);
  border-bottom: 0;
}

:host(.ic-tooltip)
  .ic-tooltip-container[data-popper-placement^="left"]
  > .ic-tooltip-arrow {
  width: var(--ic-space-xxxs);
  right: calc(-1 * var(--ic-space-1px));
}

:host(.ic-tooltip)
  .ic-tooltip-container[data-popper-placement^="left"]
  > .ic-tooltip-arrow::before {
  border-radius: 0 var(--ic-border-radius) var(--ic-border-radius) 0;
  border-left: 0;
  top: calc(-1 * var(--ic-space-1px));
}

:host(.ic-tooltip)
  .ic-tooltip-container[data-popper-placement^="right"]
  > .ic-tooltip-arrow {
  width: var(--ic-space-xxxs);
  left: calc(-1 * var(--ic-space-xxs));
}

:host(.ic-tooltip)
  .ic-tooltip-container[data-popper-placement^="right"]
  > .ic-tooltip-arrow::before {
  border-radius: var(--ic-border-radius) 0 0 var(--ic-border-radius);
  border-right: 0;
  top: calc(-1 * var(--ic-space-1px));
}

@media (forced-colors: active) {
  :host(.ic-tooltip) .ic-tooltip-container,
  :host(.ic-tooltip) .ic-tooltip-arrow::before {
    border: var(--ic-hc-border);
  }
}
