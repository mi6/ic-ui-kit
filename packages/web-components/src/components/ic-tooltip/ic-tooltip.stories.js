import { html } from "lit-html";

const defaultArgs = {
  disableHover: false,
  maxLines: 3,
  placement: "bottom",
  label: "This is a description of the button",
  theme: "inherit",
  fixedPositioning: false,
};

export default {
  title: "Web Components/Tooltip",
  component: "ic-tooltip",
};

export const Default = {
  render: (args) =>
    html`<ic-tooltip
        label="This is a description of the button"
        id="ic-tooltip-1"
        ><button aria-describedby="ic-tooltip-1">Default</button></ic-tooltip
      >
      <ic-tooltip label="This is a description of the button" id="ic-tooltip-2"
        ><ic-button aria-describedby="ic-tooltip-2">
          Default
        </ic-button></ic-tooltip
      > `,

  name: "Default",
  height: "100px",
};

export const UsingExternalMethod = {
  render: (args) =>
    html`<script>
        let display = true;
        document
          .querySelector("#clickMe")
          .addEventListener("click", handleDisplay);
        const tooltip = document.querySelector("ic-tooltip");
        function handleDisplay() {
          tooltip.displayTooltip(display, display);
          display = !display;
        }
      </script>
      <button id="clickMe">Show tooltip</button>
      <ic-tooltip
        label="This is a description of the button"
        id="ic-tooltip-test-button-default"
        ><button aria-describedby="ic-tooltip-test-button-default">
          Default
        </button></ic-tooltip
      >`,

  name: "Using external method",
  height: "100px",
};

export const DisabledHover = {
  render: (args) =>
    html`<ic-tooltip
      label="This shows 'Code' has been copied"
      id="ic-tooltip-test-button-disable-hover"
      disable-hover="true"
      ><button
        aria-labelledby="ic-tooltip-test-button-disable-hover"
        onclick="navigator.clipboard.writeText('Code')"
      >
        Show tooltip
      </button></ic-tooltip
    >`,

  name: "Disabled Hover",
  height: "100px",
};

export const TopPlacements = {
  render: (args) =>
    html` <div style="margin: 40px">
      <ic-tooltip
        placement="top-start"
        label="This is a description of the button"
        id="ic-tooltip-test-button-top-start"
      >
        <button aria-describedby="ic-tooltip-test-button-top-start">
          Top start
        </button>
      </ic-tooltip>
      <ic-tooltip
        placement="top"
        label="This is a description of the button"
        id="ic-tooltip-test-button-top"
      >
        <button aria-describedby="ic-tooltip-test-button-top">Top</button>
      </ic-tooltip>
      <ic-tooltip
        placement="top-end"
        label="This is a description of the button"
        id="ic-tooltip-test-button-top-end"
      >
        <button aria-describedby="ic-tooltip-test-button-top-end">
          Top end
        </button></ic-tooltip
      >
    </div>`,

  name: "Top placements",
};

export const BottomPlacements = {
  render: (args) =>
    html` <div style="margin: 40px">
      <ic-tooltip
        placement="bottom-start"
        label="This is a description of the button"
        id="ic-tooltip-test-button-bottom-start"
      >
        <button aria-describedby="ic-tooltip-test-button-bottom-start">
          Bottom start
        </button>
      </ic-tooltip>
      <ic-tooltip
        placement="bottom"
        label="This is a description of the button"
        id="ic-tooltip-test-button-bottom"
      >
        <button aria-describedby="ic-tooltip-test-button-bottom">Bottom</button>
      </ic-tooltip>
      <ic-tooltip
        placement="bottom-end"
        label="This is a description of the button"
        id="ic-tooltip-test-button-bottom-end"
      >
        <button aria-describedby="ic-tooltip-test-button-bottom-end">
          Bottom end
        </button></ic-tooltip
      >
    </div>`,

  name: "Bottom placements",
};

export const RightPlacements = {
  render: (args) =>
    html` <div style="margin: 40px">
      <ic-tooltip
        placement="right-start"
        label="This is a description of the button"
        id="ic-tooltip-test-button-right-start"
      >
        <button aria-describedby="ic-tooltip-test-button-right-start">
          Right start
        </button>
      </ic-tooltip>
      <ic-tooltip
        placement="right"
        label="This is a description of the button"
        id="ic-tooltip-test-button-right"
      >
        <button aria-describedby="ic-tooltip-test-button-right">Right</button>
      </ic-tooltip>
      <ic-tooltip
        placement="right-end"
        label="This is a description of the button"
        id="ic-tooltip-test-button-right-end"
      >
        <button aria-describedby="ic-tooltip-test-button-right-end">
          Right end
        </button></ic-tooltip
      >
    </div>`,

  name: "Right placements",
};

export const LeftPlacements = {
  render: (args) =>
    html` <div style="margin-left: 400px">
      <ic-tooltip
        placement="left-start"
        label="This is a description of the button"
        id="ic-tooltip-test-button-left-start"
      >
        <button aria-describedby="ic-tooltip-test-button-left-start">
          Left start
        </button>
      </ic-tooltip>
      <ic-tooltip
        placement="left"
        label="This is a description of the button"
        id="ic-tooltip-test-button-left"
      >
        <button aria-describedby="ic-tooltip-test-button-left">Left</button>
      </ic-tooltip>
      <ic-tooltip
        placement="left-end"
        label="This is a description of the button"
        id="ic-tooltip-test-button-left-end"
      >
        <button aria-describedby="ic-tooltip-test-button-left-end">
          Left end
        </button></ic-tooltip
      >
    </div>`,

  name: "Left placements",
};

export const Truncation = {
  render: (args) =>
    html`<ic-tooltip
      label="This is a body of text that is truncated to three lines within a tooltip. It has been truncated based on the max-lines property. The number of lines in the text is clamped to the number passed through the max-lines property."
      max-lines="3"
      id="ic-tooltip-test-button-truncation"
      ><button aria-describedby="ic-tooltip-test-button-truncation">
        Default
      </button></ic-tooltip
    >`,

  name: "Truncation",
  height: "100px",
};

export const DarkBackground = {
  render: (args) =>
    html`<div style="height: 100px;background-color:#333333;">
      <ic-tooltip
        label="This is a description of the button"
        id="ic-tooltip-test-button-dark"
      >
        <button aria-describedby="ic-tooltip-test-button-dark">Default</button>
      </ic-tooltip>
    </div>`,

  name: "Dark background",
};

export const ChangingLabel = {
  render: (args) =>
    html`<script>
        var icButton = document.querySelector("#testbutton");
        function handleClick() {
          icButton.ariaLabel = "really long message";
        }
      </script>
      <div style="display: flex;justify-content:flex-end;">
        <ic-button
          variant="icon-primary"
          aria-label="refresh"
          id="testbutton"
          onclick="handleClick()"
        >
          <svg
            xmlns="http://www.w3.org/2000/svg"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            fill="#000000"
          >
            <path d="M0 0h24v24H0V0z" fill="none" />
            <path
              d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
            />
          </svg>
        </ic-button>
      </div> `,

  name: "Changing label",
};

export const PositioningStrategy = {
  render: (args) =>
    html` <script>
        var switchEl = document.querySelector("ic-switch");
        var tooltipEl = document.querySelector("ic-tooltip");
        switchEl.addEventListener("icChange", (ev) => {
          tooltipEl.fixedPositioning = ev.detail.checked;
        });
      </script>
      <div style="display:flex; flex-direction:column; gap:var(--ic-space-xs)">
        <ic-typography>
          Toggle the fixed positioning switch and hover over the button to see
          it take effect
        </ic-typography>
        <div
          style="display:flex; flex-direction:column; gap:var(--ic-space-xs); padding:var(--ic-space-md); border:1px solid var(--ic-color-text-primary); overflow:hidden; width:100px; position:relative"
        >
          <ic-tooltip
            label="This is a tooltip with a long label"
            id="ic-tooltip-test-button-1"
          >
            <ic-button aria-describedby="ic-tooltip-test-button-1">
              Button
            </ic-button>
          </ic-tooltip>
        </div>
        <ic-switch label="Fixed tooltip positioning"></ic-switch>
      </div>`,

  name: "Positioning strategy",
};

export const Playground = {
  render: (args) =>
    html`<div style="margin: 250px">
      <ic-tooltip
        label=${args.label}
        placement=${args.placement}
        max-lines=${args.maxLines}
        disable-hover=${args.disableHover}
        id="ic-tooltip-1"
        theme=${args.theme}
        fixed-positioning=${args.fixedPositioning}
      >
        <button aria-describedby="ic-tooltip-1">Native</button>
      </ic-tooltip>
      <ic-tooltip
        label=${args.label}
        placement=${args.placement}
        max-lines=${args.maxLines}
        disable-hover=${args.disableHover}
        id="ic-tooltip-2"
        theme=${args.theme}
        fixed-positioning=${args.fixedPositioning}
      >
        <ic-button aria-describedby="ic-tooltip-2"> ICDS </ic-button>
      </ic-tooltip>
    </div> `,

  args: defaultArgs,

  argTypes: {
    placement: {
      options: [
        "bottom",
        "top-start",
        "top",
        "top-end",
        "right-start",
        "right",
        "right-end",
        "bottom-end",
        "bottom-start",
        "left-end",
        "left",
        "left-start",
      ],

      control: {
        type: "select",
      },
    },

    theme: {
      options: ["inherit", "light", "dark"],

      control: {
        type: "inline-radio",
      },
    },
  },

  name: "Playground",
};
