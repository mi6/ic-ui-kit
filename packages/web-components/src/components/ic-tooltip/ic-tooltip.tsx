import {
  Component,
  Element,
  Host,
  Prop,
  h,
  Watch,
  Method,
} from "@stencil/core";
import { Instance, createPopper } from "@popperjs/core";
import { IcTooltipPlacements } from "./ic-tooltip.types";
import { onComponentRequiredPropUndefined } from "../../utils/helpers";

@Component({
  tag: "ic-tooltip",
  styleUrl: "ic-tooltip.css",
  shadow: true,
})
export class Tooltip {
  private arrow: HTMLDivElement;
  private delayedHideEvents = ["mouseleave"];
  private instantHideEvents = ["focusout"];
  private mouseOverTool: boolean = false;
  private persistTooltip = false;
  private popperInstance: Instance;
  private screenReaderOnlyStyles = {
    position: "absolute",
    left: "-10000px",
    top: "auto",
    width: "1px",
    height: "1px",
    overflow: "hidden",
  };
  private showEvents = this.disableHover
    ? ["click"]
    : ["mouseenter", "focusin"];
  private toolTip: HTMLDivElement;

  @Element() el: HTMLIcTooltipElement;

  /**
   * If `true`, the tooltip will not be displayed on hover, it will require a click.
   */
  @Prop() disableHover?: boolean = false;

  /**
   * The position of the tool-tip in relation to the parent element.
   */
  @Prop() placement?: IcTooltipPlacements = "bottom";

  /**
   * The ID of the element the tooltip is describing - for when aria-labelledby or aria-describedby is used.
   */
  @Prop({ reflect: true }) target?: string;

  /**
   * The text to display on the tooltip.
   */
  @Prop() label!: string;

  @Watch("label")
  updateLabel(newValue: string): void {
    const describedBySpan = this.el.previousElementSibling as HTMLElement;
    if (describedBySpan !== null) {
      describedBySpan.innerText = newValue;
    }
  }

  disconnectedCallback(): void {
    this.manageEventListeners("remove");
    this.popperInstance.destroy();
  }

  componentDidLoad(): void {
    this.popperInstance = createPopper(this.el, this.toolTip, {
      placement: this.placement,
      modifiers: [
        {
          name: "offset",
          options: {
            offset: [0, 10],
          },
        },
        {
          name: "arrow",
          options: {
            element: this.arrow,
          },
        },
        {
          name: "eventListeners",
          options: { scroll: false, resize: false },
        },
      ],
    });

    this.manageEventListeners("add");

    onComponentRequiredPropUndefined(
      [{ prop: this.label, propName: "label" }],
      "Tooltip"
    );

    if (this.target !== undefined) {
      const ariaDescribedBy = document.createElement("span");
      ariaDescribedBy.id = `ic-tooltip-${this.target}`;
      ariaDescribedBy.innerText = this.label;
      ariaDescribedBy.classList.add("ic-tooltip-label");
      Object.assign(ariaDescribedBy.style, this.screenReaderOnlyStyles);

      this.el.insertAdjacentElement("beforebegin", ariaDescribedBy);
    }
  }

  /**
   * Method to programmatically show/hide the tooltip without needing to interact with an anchor element
   * @param show Whether to show or hide the tooltip
   * @param persistTooltip Whether the tooltip should stay on the screen when actions are performed that would previously dismiss the tooltip, such as on hover
   */
  @Method()
  async displayTooltip(show: boolean, persistTooltip?: boolean): Promise<void> {
    this.persistTooltip = persistTooltip;
    show ? this.show() : this.hide();
  }

  private show = () => {
    this.toolTip.setAttribute("data-show", "");
    this.popperInstance.update();
  };

  private hide = () => {
    this.toolTip.removeAttribute("data-show");
    this.persistTooltip = false;
  };

  private checkCloseTooltip = () => {
    setTimeout(() => {
      if (!this.mouseOverTool && !this.persistTooltip) {
        this.hide();
      }
    }, 100);
  };

  private mouseEnterTooltip = () => {
    this.mouseOverTool = true;
  };

  private mouseLeaveTooltip = () => {
    this.mouseOverTool = false;
    this.checkCloseTooltip();
  };

  private handleKeyDown = (event: KeyboardEvent) => {
    if (event.key === "Escape" && !this.persistTooltip) {
      this.hide();
    }
  };

  private manageEventListeners = (action: "add" | "remove") => {
    const method =
      action === "add" ? "addEventListener" : "removeEventListener";

    this.showEvents.forEach((event) => {
      this.el[method](event, this.show);
      if (this.toolTip !== undefined) {
        this.toolTip[method](event, this.mouseEnterTooltip);
      }
    });

    if (!this.persistTooltip) {
      this.instantHideEvents.forEach((event) => {
        this.el[method](event, this.hide);
      });
    }

    this.delayedHideEvents.forEach((event) => {
      this.el[method](event, this.checkCloseTooltip);
      if (this.toolTip !== undefined) {
        this.toolTip[method](event, this.mouseLeaveTooltip);
      }
    });

    document[method]("keydown", this.handleKeyDown);
  };

  render() {
    const { label } = this;
    return (
      <Host class={{ "ic-tooltip": true }}>
        <div
          ref={(el) => (this.toolTip = el as HTMLDivElement)}
          role="tooltip"
          class="ic-tooltip-container"
        >
          <ic-typography variant="caption">{label}</ic-typography>
          <div
            ref={(el) => (this.arrow = el as HTMLDivElement)}
            class="ic-tooltip-arrow"
            data-popper-arrow
          ></div>
        </div>
        <slot></slot>
      </Host>
    );
  }
}
