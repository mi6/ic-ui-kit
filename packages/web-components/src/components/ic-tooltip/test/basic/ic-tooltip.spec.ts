import { newSpecPage } from "@stencil/core/testing";
import { waitForTimeout } from "../../../../testspec.setup";
import { Tooltip } from "../../ic-tooltip";

beforeAll(() => {
  jest.spyOn(console, "error").mockImplementation(jest.fn());
});

describe("ic-loading-indicator component", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip label="tooltip"></ic-tooltip>`,
    });

    expect(page.root).toMatchSnapshot(`ic-tooltip-render`);
  });

  it("should appear when triggered", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip target="test-button" label="tooltip"><button id="test-button">Click</button></ic-tooltip>`,
    });

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;

    await page.rootInstance.show(page.rootInstance.popperInstance);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBe("");
  });

  it("should appear when triggered by click", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip target="test-button" label="tooltip" disable-hover=true><button id="test-button">Click</button></ic-tooltip>`,
    });

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;

    await page.rootInstance.show(page.rootInstance.popperInstance);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBe("");
  });

  it("should disappear when triggered", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip target="test-button" label="tooltip"><button id="test-button">Click</button></ic-tooltip>`,
    });

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;

    await page.rootInstance.show(page.rootInstance.popperInstance);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBe("");

    await page.rootInstance.hide(page.rootInstance.popperInstance);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;
  });

  it("should disappear on mouseleave", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip target="test-button" label="tooltip"><button id="test-button">Click</button></ic-tooltip>`,
    });

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;

    await page.rootInstance.show(page.rootInstance.popperInstance);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBe("");

    await page.rootInstance.mouseLeaveTooltip(page.rootInstance.popperInstance);

    await waitForTimeout(200);

    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;
  });

  it("should remain when mouse remains on tooltip", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip target="test-button" label="tooltip"><button id="test-button">Click</button></ic-tooltip>`,
    });

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;

    await page.rootInstance.show(page.rootInstance.popperInstance);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBe("");

    await page.rootInstance.mouseEnterTooltip();
    await page.waitForChanges();

    await page.rootInstance.checkCloseTooltip(page.rootInstance.popperInstance);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBe("");
  });

  it("should close the tooltip on Escape keydown", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip target="test-button" label="tooltip"><button id="test-button">Click</button></ic-tooltip>`,
    });

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;

    await page.rootInstance.show(page.rootInstance.popperInstance);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBe("");

    const body = window.document.body;

    body.dispatchEvent(
      new window.window.KeyboardEvent("keydown", {
        key: "Escape",
        bubbles: true,
        cancelable: true,
      })
    );

    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;
  });

  it("should watch for label updates", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip target="test-button" label="tooltip"><button id="test-button">Click</button></ic-tooltip>`,
    });

    expect(page.rootInstance.label).toBe("tooltip");

    page.root.label = "new tooltip";

    page.rootInstance.updateLabel();
    await page.waitForChanges();

    expect(page.rootInstance.label).toBe("new tooltip");

    page.rootInstance.disconnectedCallback();
  });

  it("should open the tooltip using the displayTooltip method and should persist when persistTooltip is true", async () => {
    const page = await newSpecPage({
      components: [Tooltip],
      html: `<ic-tooltip target="test-button" label="tooltip"><button id="test-button">Click</button></ic-tooltip>`,
    });

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;

    await page.rootInstance.displayTooltip(true, true);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).not.toBeNull;

    window.document.body.dispatchEvent(
      new window.window.KeyboardEvent("keydown", {
        key: "Escape",
        bubbles: true,
        cancelable: true,
      })
    );

    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).not.toBeNull;

    await page.rootInstance.displayTooltip(false, false);
    await page.waitForChanges();

    expect(page.rootInstance.toolTip.getAttribute("data-show")).toBeNull;
  });
});
