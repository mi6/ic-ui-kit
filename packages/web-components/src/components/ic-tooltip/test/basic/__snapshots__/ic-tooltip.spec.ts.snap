// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-tooltip component getTooltipTranslate should update for bottom 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: 0px; --tooltip-translate-y: -100px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for bottom-end 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="bottom-end" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: -40px; --tooltip-translate-y: -100px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for bottom-start 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="bottom-start" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: 0px; --tooltip-translate-y: -100px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for left 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="left" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: -40px; --tooltip-translate-y: -100px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for left-end 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="left-end" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: -40px; --tooltip-translate-y: -132px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for right 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="right" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: 0px; --tooltip-translate-y: -100px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for right-end 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="right-end" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: 0px; --tooltip-translate-y: -132px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for top 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="top" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: 0px; --tooltip-translate-y: -132px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for top-end 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="top-end" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: -40px; --tooltip-translate-y: -132px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update for top-start 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="top-start" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: 0px; --tooltip-translate-y: -132px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component getTooltipTranslate should update when tooltip is outside of dialog 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip" placement="left" target="test-button">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0; --tooltip-translate-x: 0px; --tooltip-translate-y: -100px;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
  <button id="test-button">
    Click
  </button>
</ic-tooltip>
`;

exports[`ic-tooltip component should render: ic-tooltip-render 1`] = `
<ic-tooltip class="ic-tooltip" label="tooltip">
  <mock:shadow-root>
    <div class="ic-tooltip-container" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
      <ic-typography variant="caption">
        tooltip
      </ic-typography>
      <div class="ic-tooltip-arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
    <slot></slot>
  </mock:shadow-root>
</ic-tooltip>
`;
